[plug-in][java][preferences] Add files.exclude preference

svor opened the pull request:
Signed-off-by: Valeriy Svydenko <vsvydenk@redhat.com>

**The problem**
VS Code has `files.exclude` preference and some of VS Code extensions want to change value of this preference. In my case it is java-vscode extension. The last version of java-vscode extension  (0.38.0)  asks user to exclude some types of files (`.project`, `.classpath`) to hide them in **Files** view. To do that the extension wants to extend `files.exclude` preference, but Theia doesn't have such preference.

**Changes**
This PR offers to add `files.exclude` and makes it possible to hide excluded files/folders in **Files** view.

**Related issue**
https://github.com/theia-ide/theia/issues/4254

**Short demo**
![files_exclude](https://user-images.githubusercontent.com/1271546/52404790-88e94b00-2ad2-11e9-89ae-652995ed1bcf.gif)

<!-- Please provide a clear and meaningful description to the CHANGELOG.md file if this PR contributes some significant changes -->

kittaakos commented:
> This PR offers to add `files.exclude` and makes it possible to hide excluded files/folders in **Files** view.

I do not know much about the VS Code extensions but why do not we have one preference for this in Theia instead of having both the `navigator.exclude` and the new `files.exclude`?

svor commented:
@kittaakos for me it makes sense to remove `navigator.exclude` and just use `files.exclude`, but I wasn't sure of it, maybe some of Theia extensions or plugins already use `navigator.exclude` preference.

kittaakos commented:
> but I wasn't sure of it, maybe some of Theia extensions or plugins already use `navigator.exclude` preference.

Figuring this out should be part of this PR, I think. If possible, the two preferences should be merged. Otherwise, it could be confusing later why do we have two.

// cc @akosyakov

akosyakov commented:
@svor @kittaakos If `navigator.exclude` and `files.exclude` is the same thing, then let's break and rename it, please update the change log about it

svor commented:
@akosyakov @kittaakos I've made changes to use just `files.exclude` preference and deleted `navigator.eclude`. Also I've added an info about this into change log. 
The PR is ready to review, please take another look at it.

kittaakos commented on packages/navigator/src/browser/navigator-filter.ts:
```diff
@@ -60,14 +60,10 @@ export class FileNavigatorFilter {
             this.filterPredicate = this.createFilterPredicate(newValue as FileNavigatorFilter.Exclusions | undefined || {});
-            hasChanged = true;
-        }
-        if (hasChanged) {
```
We do not want to refresh the tree on all filesystem preference changes. See here: https://github.com/theia-ide/theia/blob/ba4c62a63bd5a3c3d89e97ca6ee36a2c37da75fd/packages/navigator/src/browser/navigator-tree.ts#L31

In other words, when I change `files.watcherExclude`, I guess we do not need to refresh the navigator. Did you spot a bug?

svor replied:
If you change `files.watcherExclude`, it won't do refresh the navigator because in line 65 I check if `files.exclude` preference was  changed.
`if (preferenceName === FILES_EXCLUDE_PREFERENCE) `
@kittaakos maybe I misunderstood you

kittaakos replied:
No, I was just blind. Sorry for the noise.

tsmaeder commented:
There is a list of default settings here: https://code.visualstudio.com/docs/getstarted/settings

akosyakov commented on packages/navigator/src/browser/navigator-filter.ts:
```diff
@@ -60,14 +60,10 @@ export class FileNavigatorFilter {
+    protected onFilesPreferenceChanged(event: PreferenceChangeEvent<FileSystemConfiguration>): void {
         const { preferenceName, newValue } = event;
-        if (preferenceName === 'navigator.exclude') {
+        if (preferenceName === FILES_EXCLUDE_PREFERENCE) {
```
fyi: you don't need constants, it is statically typed here, try to use another property, ts won't let you

svor replied:
@akosyakov Thanks! I see what you mean, but I use `files.exclude` in 3 places, that's why I've decided to use constant. For me it is better to change value in one place instead of doing it in 3 different places.
I've changed it a bit:
`const FILES_EXCLUDE_PREFERENCE: keyof FileSystemConfiguration = 'files.exclude';`

kittaakos commented:
Doing the review...

kittaakos commented on packages/navigator/src/browser/navigator-filter.ts:
```diff
@@ -19,29 +19,29 @@ import { Minimatch } from 'minimatch';
 
     constructor(
-        @inject(FileNavigatorPreferences) protected readonly preferences: FileNavigatorPreferences
+        @inject(FileSystemPreferences) protected readonly filesPreferences: FileSystemPreferences
```
You should leave the constructor as it was
```js
@inject(FileNavigatorPreferences) protected readonly preferences: FileNavigatorPreferences
```

and use property injection for the `FileSystemPreferences`. Otherwise, you break another API.

svor replied:
@kittaakos Thank you for your review. Can you explain more which API it can break?
I don't see any reason to inject `FileNavigatorPreferences` here it is never used in this class

kittaakos replied:
> Can you explain more which API it can break?

Sure. Subclasses must call super in the ctors. If there is any subclass of the `FileNavigatorFilter`, you will break client code if we publish a new `latest` version. If clear, please adjust , if no, or I have overlooked something, let me know.

svor replied:
@kittaakos Thank you for the explanation, it is clear! I've fixed that.

kittaakos commented:
I'll give it a try once more, but it should be fine...

kittaakos commented on packages/navigator/src/browser/navigator-filter.ts:
```diff
@@ -60,14 +64,10 @@ export class FileNavigatorFilter {
 
-    protected onPreferenceChanged(event: PreferenceChangeEvent<FileNavigatorConfiguration>): void {
-        let hasChanged = false;
+    protected onFilesPreferenceChanged(event: PreferenceChangeEvent<FileSystemConfiguration>): void {
```
This cannot be renamed. Please restore the original method name. Sorry for not mentioning it before :(

svor replied:
I left this method as it is, and added 
`protected onPreferenceChanged(event: PreferenceChangeEvent<FileNavigatorConfiguration>): void` with empty body. Please take another look at it :)
